petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+
+        if (method_scheme_raw := kwargs.pop(METHOD_SCHEME, None)) is not None:
+            method_scheme = {
+                (
+                    tuple(
+                        [
+                            Method(method_str)
+                            for method_str in definition[PREVIOUS_METHODS]
+                        ]
+                    )
+                    if definition[PREVIOUS_METHODS] is not None
+                    else None
+                ): definition[NEXT_METHOD]
+                for definition in method_scheme_raw
+            }
+            kwargs[METHOD_SCHEME] = method_scheme
+
+        return super().read_arguments_from_yaml_dict(yaml_dict=kwargs)
+
+    def update_after_calibration(
+        self,
+        history: Dict[str, Model],
+        local_history: Dict[str, Model],
+        criterion: Criterion,
+    ) -> None:
+        """See `CandidateSpace.update_after_calibration`."""
+        # In case we jumped to most distant in the last iteration,
+        # here we reset the jumped variable to False
+        self.jumped_to_most_distant = False
+
+        self.history = history
+
+        if not self.update_from_local_history(
+            local_history=local_history, criterion=criterion
+        ):
+            logging.info("Switching method")
+            self.switch_method()
+            self.switch_inner_candidate_space(history)
+            logging.info(
+                "Method switched to ", self.inner_candidate_space.method
+            )
+
+        self.method_history.append(self.method)
+        return self.jumped_to_most_distant
+
+    def update_from_local_history(
+        self, local_history: Dict[str, Model], criterion: Criterion
+    ) -> bool:
+        """Update the self.best_models with the latest local_history
+        and determine if there was a new best model. If so, return
+        True. False otherwise."""
+
+        found_new_best = False
+        for model_id in local_history:
+            if (
+                self.best_model_of_current_run == VIRTUAL_INITIAL_MODEL
+                or default_compare(
+                    self.best_model_of_current_run,
+                    local_history[model_id],
+                    criterion,
+                )
+            ):
+                found_new_best = True
+                self.best_model_of_current_run = local_history[model_id]
+
+            if len(self.best_models) < self.most_distant_max_number:
+                self.insert_model_into_best_models(
+                    model_to_insert=local_history[model_id],
+                    criterion=criterion,
+                )
+            elif default_compare(
+                self.best_models[self.most_distant_max_number - 1],
+                local_history[model_id],
+                criterion,
+            ):
+                self.insert_model_into_best_models(
+                    model_to_insert=local_history[model_id],
+                    criterion=criterion,
+                )
```

Comment:
These conditions have the same outcome, could combine

Reply:
I agree, I'll push it now